### ISSUE TITLE
Forced user verification on Ubikeys not supported

### ISSUE THREAD

**Issue by FelixOliverLange** [nextcloud/twofactor_webauthn#599]
### Steps to reproduce
1. on a UbiKey, enforce user verification to always be required for FIDO2 actions like signin: `ykman fido config toggle-always-uv` (note: this requires yubikey-manager >= 5.3.0 and can be done before or after enrolling via this plugin - it doesn't matter)
2. attempt to sign-in to nextcloud using that ubikey (which won't work, see below)
3. to validate that the key is in fact working, toggle user verification enforcement again (same command) so that it's off and re-attempt step 2 (which will now succeed) 

### Expected behaviour
Login should work with enforced user verification. Note: During enrollment, this is already working correctly - just not during login

### Actual behaviour
The browser raises a notification to touch the ubikey. However, no popup is presented to request user verification and the LED of the key also does not indicate that any action is required. Touching it also does nothing, as the ubikey expects verification to be done but that's not possible as no popup for user valdiation is presented. Eventually the login fails (I presume a timeout?) and so login with enforced verification doesn't work. I assume this to be because this extension marks user verification to be discouraged, but I'm not certain about that.

I'd note that as the key is only used as second factor, someone who obtains the key would not directly gain access to nextcloud. From that perspective, it can be argued that enforced verification is unnecessary or overkill. However, this setting of ubikey is only available for the entire device. So as soon as other credentials are on it that would give instant access it makes sense to enforce user verification (as those other services may choose to not enforce it on their end). But if done, the user is no longer able to use it for this integration. A workaround is to always toggle validation on and off before / after login, but that's quite cumbersome and dangerous, as users may forget to toggle it back on after login.

### Server configuration

**Operating system**:
hosted environment (I presume debian)

**Web server:**
Apache (hosted, so I'm not sure)

**Database:**
MySQL 28.0.5.1

**PHP version:**
8.1

**Version:** (see admin page)
1.4.0

**Updated from an older version or fresh install:**
Fresh install of the extension

**List of activated apps:**
  - activity: 2.20.0
  - admin_audit: 1.18.0
  - bruteforcesettings: 2.8.0
  - calendar: 4.7.6
  - circles: 28.0.0
  - cloud_federation_api: 1.11.0
  - comments: 1.18.0
  - contactsinteraction: 1.9.0
  - dashboard: 7.8.0
  - dav: 1.29.1
  - deck: 1.12.2
  - encryption: 2.16.0
  - federatedfilesharing: 1.18.0
  - federation: 1.18.0
  - files: 2.0.0
  - files_external: 1.20.0
  - files_pdfviewer: 2.9.0
  - files_reminders: 1.1.0
  - files_sharing: 1.20.0
  - files_trashbin: 1.18.0
  - files_versions: 1.21.0
  - firstrunwizard: 2.17.0
  - logreader: 2.13.0
  - lookup_server_connector: 1.16.0
  - nextcloud_announcements: 1.17.0
  - notes: 4.10.0
  - notifications: 2.16.0
  - oauth2: 1.16.3
  - password_policy: 1.18.0
  - photos: 2.4.0
  - privacy: 1.12.0
  - provisioning_api: 1.18.0
  - recommendations: 2.0.0
  - related_resources: 1.3.0
  - serverinfo: 1.18.0
  - settings: 1.10.1
  - sharebymail: 1.18.0
  - support: 1.11.1
  - survey_client: 1.16.0
  - systemtags: 1.18.0
  - tasks: 0.16.0
  - text: 3.9.1
  - theming: 2.3.0
  - twofactor_backupcodes: 1.17.0
  - twofactor_nextcloud_notification: 3.9.0
  - twofactor_totp: 10.0.0-beta.2
  - twofactor_webauthn: 1.4.0
  - updatenotification: 1.18.0
  - user_status: 1.8.1
  - viewer: 2.2.0
  - weather_status: 1.8.0
  - workflowengine: 2.10.0

**The content of config/config.php:**
"instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "28.0.5.1",
        "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "maintenance": false,
        "theme": "",
        "loglevel": 0,
        "updater.release.channel": "stable",
        "encryption.legacy_format_support": true,
        "encryption.key_storage_migrated": false,
        "default_phone_region": "***REMOVED SENSITIVE VALUE***",
        "updater.secret": "***REMOVED SENSITIVE VALUE***"

#### Client configuration
**Browser:**
115.12.0esr 

**Operating system:**
OpenSuse Leap 15.6

##### Web server error log
(pretty huge, let me know if you really need it and I'll try to create a minimal log of the event)

##### Server log (data/nextcloud.log)
(pretty huge, let me know if you really need it and I'll try to create a minimal log of the event)

##### Browser log
JS console log below:

The resource from “***REMOVED SENSITIVE VALUE***/index.php/login/selectchallenge?redirect_url=/index.php/js/core/merged-template-prepend.js?v%3D93c3f87b-5” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). webauthn

Loading failed for the <script> with source “***REMOVED SENSITIVE VALUE***/index.php/js/core/merged-template-prepend.js?v=93c3f87b-5”. webauthn:24:208

[DEBUG] twofactor_webauthn: start sign Object { app: "twofactor_webauthn", uid: "***REMOVED SENSITIVE VALUE***", level: 0 } ConsoleLogger.js:65:18

[DEBUG] twofactor_webauthn: Starting webauthn authentication 
Object { 
publicKey: Object { 
allowCredentials: Array [ {…} ]
challenge: Uint8Array(32) [ … ]
extensions: Object { loc: true, appid: "***REMOVED SENSITIVE VALUE***" }
rpId: "***REMOVED SENSITIVE VALUE***"
timeout: 60000
userVerification: "discouraged"
uid: "***REMOVED SENSITIVE VALUE***"
}
ConsoleLogger.js:65:18

[ERROR] twofactor_webauthn: challenge failed 
Object { app: "twofactor_webauthn", uid: "***REMOVED SENSITIVE VALUE***", level: 0, error: DOMException }
ConsoleLogger.js:74:18

**Comment by st3iny** [comment]
Thank you for the report. 

We indeed discourage user verification when registering and logging in. A simple fix would be to prefer but not require user verification.

Unfortunately, my YubiKeys do not support the `toggle-always-uv` config so I cannot test my patch. Would you be willing to apply the following patch to your instance and see if your key works again?

https://patch-diff.githubusercontent.com/raw/nextcloud/twofactor_webauthn/pull/601.patch

**Comment by st3iny** [comment]
Thank you for the report. 

We indeed discourage user verification when registering and logging in. A simple fix would be to prefer but not require user verification.

Unfortunately, my YubiKeys do not support the `toggle-always-uv` config so I cannot test my patch. Would you be willing to apply the following patch to your instance and see if your key works again?

https://patch-diff.githubusercontent.com/raw/nextcloud/twofactor_webauthn/pull/602.patch

---

**EDIT:** Fixed wrong patch URL.

**Comment by FelixOliverLange** [comment]
Thank you very much for the quick patch! I will try to set up a fresh test instance of nextcloud 28 this weekend to test it, just to avoid any negative side effects on my instance and eliminate potential compatibility issues against the patch.

**Comment by FelixOliverLange** [comment]
After some more testing, I've found newer versions of Firefox on Linux to work with the enforced user verification and this app, even without the patch that @st3iny thankfully provided. **Specifically, Firefox 127.0.2 works, whereas 115 ESR does not.** As my distribution only offers ESR I've switched to the [Flatpak version shipped from Mozilla](https://support.mozilla.org/en-US/kb/install-firefox-linux#w_install-from-flatpak), which did the trick. 

Interestingly, the Passwordless option of Nextcloud works with enforced user verification on Firefox 115 ESR without any issue, there the input dialog is shown. Only the second factor then doesn't work as it again does not prompt for a pin. I'm not sure how different the codebases are, but I'm a bit surprised by this behavior, as that passwordless login also does not create a FIDO2 Passkey that I'd be able to see in the Yubikey manager - just like this app. 

In either case, this means that this issue is isolated to "older" versions of Firefox, with newer versions and Chromium both working fine. Given that Firefox 115 ESR [will only be maintained for another two months](https://endoflife.date/firefox),  I'll go ahead and close this issue. Thank you very much for the patch @st3iny (which didn't turn out to be required) & sorry for the additional maintenance efforts.
